airbyte-cdk/java/airbyte-cdk/s3-destinations/src/main/kotlin/io/airbyte/cdk/integrations/destination/s3/jsonschema/AirbyteJsonSchemaType.kt

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.cdk.integrations.destination.s3.jsonschema
+
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.node.ArrayNode
+import com.fasterxml.jackson.databind.node.ObjectNode
+
+enum class AirbyteJsonSchemaType {
+    BOOLEAN,
+    INTEGER,
+    NUMBER,
+    STRING,
+    BINARY_DATA,
+    DATE,
+    TIMESTAMP_WITH_TIMEZONE,
+    TIMESTAMP_WITHOUT_TIMEZONE,
+    TIME_WITH_TIMEZONE,
+    TIME_WITHOUT_TIMEZONE,
+    ARRAY_WITHOUT_ITEMS,
+    ARRAY_WITH_ITEM,
+    ARRAY_WITH_ITEMS,
+    OBJECT_WITHOUT_PROPERTIES,
+    OBJECT_WITH_PROPERTIES,
+    UNION,
+    COMBINED;
+
+    fun matchesValue(tree: JsonNode): Boolean {
+        return when (this) {
+            BOOLEAN -> tree.isBoolean
+            INTEGER -> tree.isIntegralNumber || tree.isInt || tree.isBigInteger
+            NUMBER ->
+                tree.isDouble ||
+                    tree.isFloat ||
+                    tree.isIntegralNumber ||
+                    tree.isInt ||
+                    tree.isBigInteger
+            STRING -> tree.isTextual
+            BINARY_DATA -> tree.isBinary || tree.isTextual
+            DATE,
+            TIMESTAMP_WITH_TIMEZONE,
+            TIMESTAMP_WITHOUT_TIMEZONE,
+            TIME_WITH_TIMEZONE,
+            TIME_WITHOUT_TIMEZONE -> tree.isTextual
+            ARRAY_WITHOUT_ITEMS,
+            ARRAY_WITH_ITEM,
+            ARRAY_WITH_ITEMS -> tree.isArray
+            OBJECT_WITHOUT_PROPERTIES -> tree.isObject
+            OBJECT_WITH_PROPERTIES -> tree.isObject
+            UNION,
+            COMBINED -> throw IllegalStateException("Union type cannot be matched")
```

Comment:
I think when we migrate to using a dedicated airbyte type, this distinction should go away, but right now the mapper code needs it to know where to go to get the union options. (`Union -> oneOf :: array of complete type objects`, `Combined -> type :: array of type names`)

Which is messy in practice, it would be better if there were just a bespoke AirbyteType.Union class with an `options` field.